pkg/reconciler/stats_reporter.go

```diff
@@ -0,0 +1,106 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconciler
+
+import (
+	"context"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"knative.dev/pkg/metrics"
+)
+
+const (
+	knativeServingEditCountName = "knativeserving_edit_count"
+)
+
+var (
+	knativeServingEditCountStat = stats.Int64(
+		knativeServingEditCountName,
+		"Number of KnativeServing changes",
+		stats.UnitDimensionless)
+	// Create the tag keys that will be used to add tags to our measurements.
+	// Tag keys must conform to the restrictions described in
+	// go.opencensus.io/tag/validate.go. Currently those restrictions are:
+	// - length between 1 and 255 inclusive
+	// - characters are printable US-ASCII
+	reconcilerTagKey = tag.MustNewKey("reconciler")
+	keyTagKey        = tag.MustNewKey("key")
+)
+
+func init() {
+	// Create views to see our measurements. This can return an error if
+	// a previously-registered view has the same name with a different value.
+	// View name defaults to the measure name if unspecified.
+	views := []*view.View{{
+		Description: knativeServingEditCountStat.Description(),
+		Measure:     knativeServingEditCountStat,
+		Aggregation: view.Count(),
+		TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+	}}
+
+	if err := view.Register(views...); err != nil {
+		panic(err)
+	}
+}
+
+// StatsReporter defines the interface for sending metrics
+type StatsReporter interface {
+	// ReportKnativeServingReady reports the count and latency metrics for a reconcile operation
+	ReportKnativeServingEdit(NamespacedName string) error
+}
+
+// Reporter holds cached metric objects to report metrics
+type reporter struct {
+	reconciler string
+	ctx        context.Context
+}
+
+// NewStatsReporter creates a reporter that collects and reports metrics
+func NewStatsReporter(reconciler string) (StatsReporter, error) {
+	// Reconciler tag is static. Create a context containing that and cache it.
+	ctx, err := tag.New(
+		context.Background(),
+		tag.Insert(reconcilerTagKey, reconciler))
+	if err != nil {
+		return nil, err
+	}
+	return &reporter{reconciler: reconciler, ctx: ctx}, nil
+}
+
+func(r *reporter) getContext(key string) (context.Context, error) {
+	ctx, err := tag.New(
+		context.Background(),
+		tag.Insert(reconcilerTagKey, r.reconciler),
+		tag.Insert(keyTagKey, key))
+	if err != nil {
+		return nil, err
+	}
+	return ctx, nil
+}
+
+// ReportInstallSuccess captures successful KnativeServing CR install
+func (r *reporter) ReportKnativeServingEdit(key string) error {
```

Comment:
There should be validation of the values that are passed in here. The valid values should be defined as const strings somewhere in this file that callers can re-use.